### ISSUE TITLE
Validation of YOLO pretrained COCO on custom Dataset - Zero metrics

### ISSUE THREAD

**Issue by adriengoleb** [ultralytics/ultralytics#18920]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hello, I want to test the Yolo model pre-trained on YOLO on my custom dataset.
My test dataset contains only 1 label class: the boat coco class.

My YAML file is following :  
```

path: N:/IA/data_2024/split_coco_to_data
train:   
val: images/test  
test:  


# Classes
names:
  0: person
  1: bicycle
  2: car
  3: motorcycle
  4: airplane
  5: bus
  6: train
  7: truck
  8: boat
  9: traffic light
  10: fire hydrant
  11: stop sign
  12: parking meter
  13: bench
  14: bird
  15: cat
  16: dog
  17: horse
  18: sheep
  19: cow
  20: elephant
  21: bear
  22: zebra
  23: giraffe
  24: backpack
  25: umbrella
  26: handbag
  27: tie
  28: suitcase
  29: frisbee
  30: skis
  31: snowboard
  32: sports ball
  33: kite
  34: baseball bat
  35: baseball glove
  36: skateboard
  37: surfboard
  38: tennis racket
  39: bottle
  40: wine glass
  41: cup
  42: fork
  43: knife
  44: spoon
  45: bowl
  46: banana
  47: apple
  48: sandwich
  49: orange
  50: broccoli
  51: carrot
  52: hot dog
  53: pizza
  54: donut
  55: cake
  56: chair
  57: couch
  58: potted plant
  59: bed
  60: dining table
  61: toilet
  62: tv
  63: laptop
  64: mouse
  65: remote
  66: keyboard
  67: cell phone
  68: microwave
  69: oven
  70: toaster
  71: sink
  72: refrigerator
  73: book
  74: clock
  75: vase
  76: scissors
  77: teddy bear
  78: hair drier
  79: toothbrush

```

 **As I don't have the training data for the COCO, I'm leaving the train path in the YAML file blank.**

I get these results and error when I attempt to print the metrics : 

```
[]
```

I get the test images with the predictions as well as the predictions.json file containing all the predictions.

However, I don't get any output metrics, which are set to 0 by default. I don't understand where the error comes from, given that my YAML file is well defined, as is my entire folder tree.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @adriengoleb, thank you for your interest in Ultralytics 🚀! 

It looks like you are encountering an issue with validating a YOLO pretrained model on your custom dataset where the output metrics are zero. To assist you better, we need a bit more information. If this is a 🐛 Bug Report, could you provide a minimum reproducible example (MRE) to help us debug? This should include a brief script or steps that demonstrate the issue using the smallest possible dataset or setup. 

If instead this is a custom training ❓ Question, please ensure to share as much context as possible, including:
- How your dataset is structured and whether it aligns with the expected formats.
- Partial validation/training/testing logs where relevant.
- Specific details about your setup, such as the exact command you ran or configuration changes you’ve made.

Additionally, ensure you're using the latest version of Ultralytics in a Python>=3.8 environment and have all required dependencies, including PyTorch>=1.8, properly installed. To upgrade, run the following command:

```bash
pip install -U ultralytics
```

For running YOLO, we recommend verified environments such as Jupyter Notebooks, cloud-based platforms, or local installations if all dependencies are set up correctly. Metrics output issues can sometimes arise from misaligned datasets or configurations, so double-checking your YAML file paths, class names, and data structure could also help. 

This is an automated response to guide you while an Ultralytics engineer reviews your issue soon. Feel free to add any additional details in the meantime to ensure the best support. Thank you for using Ultralytics! 😊

**Comment by Y-T-G** [comment]
I don't see the error

**Comment by adriengoleb** [comment]
It's not really an error but just metric results equal to zero which is not understable and no metrics by classes.

![Image](https://github.com/user-attachments/assets/bff65fcf-757e-4e86-ae29-9240ede7dc61)

**Comment by Y-T-G** [comment]
What are you printing? What's the code?

**Comment by adriengoleb** [comment]
here below the val script : 

```
from ultralytics import YOLO
import os

# Initialization of validation
# Load a model
model = YOLO("P:/AGO/Developpement/Models/YOLO_V8_data_TEST_DETECTION/Inference/Pretrained/coco/yolov8n.pt")



# Specify the save directory fosr training runs
save_dir = 'P:/AGO/Developpement/Models/YOLO_V8_data_TEST_DETECTION/runs/detect'
os.makedirs(save_dir, exist_ok=True)

# Use the model
if __name__ == '__main__':    
    #results = model.train(data='N:/IA/SMD_0/SMD_IR_v0/SMD_IR/nir2nir/data.yaml', batch=4, device='cuda', epochs= 100, imgsz=1920, save_dir=save_dir, project = '')

    # Validate with a custom dataset
    metrics = model.val(data='N:/IA/Mayotte_2024/split_coco_to_data/data_coco_to_data.yaml',batch=1,device='cuda', conf=0.25, epochs= 100, imgsz=1920, project=save_dir, save_json=True)
    print(metrics.maps)  # map50-95
    print(type(metrics))

```

**Comment by Y-T-G** [comment]
The metrics are empty because it either didn't detect any objects or didn't find any labels in your dataset.

**Comment by Y-T-G** [comment]
What does your confusion matrix show?

**Comment by adriengoleb** [comment]
The confusion matrix looks like this :

![Image](https://github.com/user-attachments/assets/8a1573d5-d655-4054-ada1-d516bb425205)

No ‘boat’ class is detected. These are false negatives.

On the other hand, there are a lot of (false) detections of other classes (giraffe freesbie, etc.), which are false positives. Why aren't they represented in the confusion matrix? I only have the kite and the boat classes represented ...

**Comment by Y-T-G** [comment]
Is this latest Ultralytics?

**Comment by adriengoleb** [comment]
Yes I have the latest version of Ultralytics. 
It's weird that the matrix confusion doesn't print all the false positives. 
The model predicts classes when it's background. But not all the classes are printed in the matrix.

**Comment by Y-T-G** [comment]
You can check `results.confusion_matrix.matrix`

**Comment by adriengoleb** [comment]
I have these results :

[confusion_matrix.csv](https://github.com/user-attachments/files/18579837/confusion_matrix.csv)

![Image](https://github.com/user-attachments/assets/93b87bf5-1303-48a3-9d49-a750bb1f8b90)

I have only this "756" value.

**Comment by glenn-jocher** [comment]
The value "756" suggests that the model is predicting a significant number of false positives for the background, which are not being matched to any ground truth. Ensure that your dataset is correctly labeled, including background images, and verify the `conf` and `iou_thres` parameters in your configuration. Mislabeling or incomplete annotations might be causing this issue.

**Comment by Y-T-G** [comment]
> I have these results :
> 
> [confusion_matrix.csv](https://github.com/user-attachments/files/18579837/confusion_matrix.csv)
> 
> ![Image](https://github.com/user-attachments/assets/93b87bf5-1303-48a3-9d49-a750bb1f8b90)
> 
> I have only this "756" value.
> 

The confusion matrix uses a confidence threshold of 0.25. While validation uses 0.001 which is required for mAP calculation. So you will see a lot more predictions for the validation.

**Comment by adriengoleb** [comment]
Yes, I completely understand what you're saying.

But I still have 2 questions:

- The model predicts classes when it is in the background. But not all the classes are printed in the confusion matrix (only kite and boat), why? Whereas in my json results and my ouptuts images, a greater number of classes are predicted! Airplaine and freesbe are predicted several times even though they don't appear in the confusion matrix, why?

- Why is the output of ```results.confusion_matrix.matrix``` different from the matrix automatically output in jpg?

**Comment by adriengoleb** [comment]
Also:

Testing with a default conf of 0.001 :
   ```metrics = model.val(data=‘N:/IA/Mayotte_2024/split_coco_to_Mayotte/data_coco_to_mayotte.yaml’,batch=1,device=‘cuda’,epochs= 100, imgsz=1920, project=save_dir, save_json=True)```

I get a predictions.json with more predictions.

On the other hand, I get exactly the same previous matrix and the same predictions per image. I have the impression that the conf parameter is not applied in these 2 cases : 

![Image](https://github.com/user-attachments/assets/326d5be6-fed3-4be2-b388-50ee6e9ad22a)

**Comment by Y-T-G** [comment]
> Also:
> 
> Testing with a default conf of 0.001 :
>    ```metrics = model.val(data=‘N:/IA/Mayotte_2024/split_coco_to_Mayotte/data_coco_to_mayotte.yaml’,batch=1,device=‘cuda’,epochs= 100, imgsz=1920, project=save_dir, save_json=True)```
> 
> I get a predictions.json with more predictions.
> 
> On the other hand, I get exactly the same previous matrix and the same predictions per image. I have the impression that the conf parameter is not applied in these 2 cases : 
> 
> ![Image](https://github.com/user-attachments/assets/326d5be6-fed3-4be2-b388-50ee6e9ad22a)

0.001 is the default threshold for val. The confidence matrix will still use 0.25 in that case. You can use 0.01 which will make both the validation and matrix use 0.01

**Comment by Y-T-G** [comment]
You can't use 0.001 for the matrix. It will automatically use 0.25 when 0.001 is passed because it's hardcoded to use that for the default confidence threshold. You will have to use a different threshold to change the threshold used by the matrix.

**Comment by adriengoleb** [comment]
But you say that ‘You can use 0.01, which means that the validation and the matrix will both use 0.01’.

How can we change the threshold then and ensure the consistency of the results in my case?

**Comment by Y-T-G** [comment]
You can pass `conf=0.01` to model.val

**Comment by Y-T-G** [comment]
What are you referring to by output images? The batch plots?

**Comment by Y-T-G** [comment]
The batch plots use a confidence threshold of 0.25. It's hardcoded and cannot be changed even if you pass a different conf

**Comment by adriengoleb** [comment]
> The batch plots use a confidence threshold of 0.25. It's hardcoded and cannot be changed even if you pass a different conf

1/ So if I summarise, the val mode conf parameter only influences the predict.json file predictions? For the batch plot and the confusion matrix, the confs are fixed ?

2/ The confidence threshold for batch plots cannot be lowered ok, but not even increased? Is there really no way of modifying it in the code?

**Comment by adriengoleb** [comment]
> there are a lot of (false) detections of other classes (giraffe freesbie, etc.), which are false positives. Why aren't they represented in the confusion matrix? I only have the kite and the boat classes represented ...

Also, I'm back to my original problem, in the case where I fix the conf paramater to ```conf=0.25```

There are a lot of (false) detections of other classes (giraffe freesbie, etc.) in the batch plots, which are false positives. Why aren't they represented in the confusion matrix? I only have the kite and the boat classes represented ...

**Comment by Y-T-G** [comment]
> > The batch plots use a confidence threshold of 0.25. It's hardcoded and cannot be changed even if you pass a different conf
> 
> 1/ So if I summarise, the val mode conf parameter only influences the predict.json file predictions? For the batch plot and the confusion matrix, the confs are fixed ?
> 
The conf you pass also affects the matrix as long as you pass something other than the default which is 0.001.

> 2/ The confidence threshold for batch plots cannot be lowered ok, but not even increased? Is there really no way of modifying it in the code?

Confidence threshold is lower limit. So if you increase beyond 0.25, it will also affect batch plots

**Comment by Y-T-G** [comment]
> > there are a lot of (false) detections of other classes (giraffe freesbie, etc.), which are false positives. Why aren't they represented in the confusion matrix? I only have the kite and the boat classes represented ...
> 
> Also, I'm back to my original problem, in the case where I fix the conf paramater to `conf=0.25` for the val mode.
> 
> There are a lot of (false) detections of other classes (giraffe freesbie, etc.) in the batch plots, which are false positives. Why aren't they represented in the confusion matrix? I only have the kite and the boat classes represented ...

How did you check? Did you check the index of the matrix returned for those classes?

**Comment by adriengoleb** [comment]
Thanks ! I was able to solve my original problems. My confusion matrix shows all the false positives and false negatives correctly. It's perfect.

However, for some reason, **only the first 3 box plots** are returned. It's very weird.

![Image](https://github.com/user-attachments/assets/fdcb391f-cb30-4183-856a-bb8386a61316)

```
I recall my workflow : Ultralytics 8.3.70 🚀 Python-3.12.3 torch-2.2.2+cu121 CUDA:0 (NVIDIA GeForce RTX 4070 Laptop GPU, 8188MiB)
YOLOv8n summary (fused): 168 layers, 3,151,904 parameters, 0 gradients, 8.7 GFLOPs

```

And my script for the val mode : 

```
from ultralytics import YOLO
import pandas as pd
import os

# Initialization of validation
# Load a model
model = YOLO("P:/AGO/Developpement/Models/YOLO_V8_MAYOTTE_TEST_DETECTION/Inference/Pretrained/coco/yolov8n.pt")


# Specify the save directory fosr training runs
save_dir = 'P:/AGO/Developpement/Models/YOLO_V8_MAYOTTE_TEST_DETECTION/runs/detect'
os.makedirs(save_dir, exist_ok=True)

# Use the model
if __name__ == '__main__':    

    # Validate with a custom dataset
    
    metrics = model.val(data='N:/IA/Mayotte_2024/split_coco_to_Mayotte/data_coco_to_mayotte.yaml',batch=1,device='cuda', conf=0.25, epochs= 100, imgsz=1920, project=save_dir, save_json=True, plots=True)
    
    print(metrics.maps)  # map50-95
    print(type(metrics))
    print(metrics.confusion_matrix.matrix)

````

**Comment by Y-T-G** [comment]
Ultralytics only plots the first 3 batches.

**Comment by Y-T-G** [comment]
If you want to plot FPs, TPs, FNs for all images, you can try this

https://github.com/ultralytics/ultralytics/pull/18868

**Comment by adriengoleb** [comment]
I still have the same problem with adding the Visualize parameter:

```
metrics = model.val(data='N:/IA/Mayotte_2024/split_coco_to_Mayotte/data_coco_to_mayotte.yaml',batch=1,device='cuda', conf=0.01, epochs= 100, imgsz=1920, project=save_dir, save_json=True, visualize=True)

```` 

So far I've managed to display all the predictions for a given identical image plot with batch=1.

But since I updated the version of ultralytics, this is limited to 3 images, not all the images.

How can I get back my previous results, i.e. all the predictions for an image grouped together in a single file?

**Comment by Y-T-G** [comment]
Ultralytics never supported plotting more than 3 batches. So I am not sure how you got that before.

You can use the code here to plot all batches:
https://github.com/ultralytics/ultralytics/issues/18177#issuecomment-2536456590

**Comment by adriengoleb** [comment]
> Ultralytics never supported plotting more than 3 batches. So I am not sure how you got that before.
> 
> You can use the code here to plot all batches: [#18177 (comment)](https://github.com/ultralytics/ultralytics/issues/18177#issuecomment-2536456590)

Thanks, just what I needed !

**Comment by glenn-jocher** [comment]
You're welcome! 😊 If you have any more questions as you continue working with YOLO, feel free to ask. We're here to help!
